Userland/Libraries/LibGfx/ImageFormats/TGALoader.cpp

```diff
@@ -174,55 +62,84 @@ IntSize TGAImageDecoderPlugin::size()
     return IntSize { m_context->header.width, m_context->header.height };
 }
 
-ErrorOr<void> TGAImageDecoderPlugin::decode_tga_header()
+static ErrorOr<void> ensure_header_validity(TGAHeader const& header, size_t whole_image_stream_size)
 {
-    auto& reader = m_context->reader;
-    if (reader->data().size() < sizeof(TGAHeader))
-        return Error::from_string_literal("Not enough data to be a TGA image");
-
-    m_context->header.id_length = reader->read_u8();
-    m_context->header.color_map_type = reader->read_u8();
-    m_context->header.data_type_code = static_cast<TGADataType>(reader->read_u8());
-    m_context->header.color_map_origin = reader->read_i16();
-    m_context->header.color_map_length = reader->read_i16();
-    m_context->header.color_map_depth = reader->read_u8();
-    m_context->header.x_origin = reader->read_i16();
-    m_context->header.y_origin = reader->read_i16();
-    m_context->header.width = reader->read_u16();
-    m_context->header.height = reader->read_u16();
-    m_context->header.bits_per_pixel = reader->read_u8();
-    m_context->header.image_descriptor = reader->read_u8();
-
-    auto bytes_remaining = reader->data().size() - reader->index();
-
-    if (m_context->header.data_type_code == TGADataType::UncompressedRGB && bytes_remaining < static_cast<u64>(m_context->header.width) * m_context->header.height * (m_context->header.bits_per_pixel / 8))
-        return Error::from_string_literal("Not enough data to read an image with the expected size");
-
-    if (m_context->header.bits_per_pixel < 8 || m_context->header.bits_per_pixel > 32)
+    auto bytes_remaining = whole_image_stream_size - sizeof(TGAHeader);
+    if ((header.bits_per_pixel % 8) != 0 || header.bits_per_pixel < 8 || header.bits_per_pixel > 32)
         return Error::from_string_literal("Invalid bit depth");
+    if (header.data_type_code == TGADataType::UncompressedRGB && bytes_remaining < static_cast<u64>(header.width) * header.height * (header.bits_per_pixel / 8))
+        return Error::from_string_literal("Not enough data to read an image with the expected size");
+    return {};
+}
 
+ErrorOr<void> TGAImageDecoderPlugin::decode_tga_header()
+{
+    // NOTE: Read the image size before reading the header from the stream
+    // as it will truncate that from what we need - the whole image size, including
+    // the header size in it.
+    auto image_bytes_size = static_cast<FixedMemoryStream const&>(m_context->stream).bytes().size();
```

Comment:
Can you cherry-pick https://github.com/SerenityOS/serenity/commit/afc2f5d526cbabf88794a602c6a60aa3d4fe6248?
This `static_cast` really bugs me